opentrons-ai-server/deploy.py

```
@@ -0,0 +1,149 @@
+        else:
+            raise ValueError("Invalid environment specified")
+
+
+class Deployment:
+    def __init__(self, env: str) -> None:
+        self.config: DeploymentConfig = DeploymentConfig()
+        self.env: str = env
+        self.s3_key: str = self.config.S3_KEY
+        self.s3_zip_path: Path = self.config.S3_ZIP_PATH
+        self.s3_bucket = self.config.get_bucket(env)
+        self.function_name = self.config.get_function_name(env)
+        self.lambda_client = boto3.client("lambda")
+        self.s3_client = boto3.client("s3")
+
+    def upload_to_s3(self) -> None:
+        """Upload the packaged Lambda function to S3."""
+        print(f"Uploading to S3 bucket {self.s3_bucket} with key {self.s3_key}")
+        try:
+            self.s3_client.upload_file("function.zip", self.s3_bucket, self.s3_key)
+            print("Uploaded to S3 successfully!")
+        except NoCredentialsError:
+            print("Credentials not available")
+        except PartialCredentialsError:
+            print("Incomplete credentials configuration")
+
+    def update_lambda(self) -> None | str:
+        """Update a Lambda function using the uploaded S3 object."""
+        print(f"Updating Lambda function: {self.function_name}")
+        print("If the code has not changed in the S3, the version will not be updated.")
+        try:
+            # Try to update the existing function
+            response = self.lambda_client.update_function_code(
+                FunctionName=self.function_name, S3Bucket=self.s3_bucket, S3Key=self.s3_key, Publish=True
+            )
+            print("Updated Lambda function:")
+            print(response)
+            version = str(response["Version"])
+            print(f"New version: {version}")
+            return version
+        except Exception as e:
+            print(f"Failed to update Lambda function: {e}")
+        return None
+
+    def wait_for_lambda_status(self, version: str) -> None:
```

this function really needs to indicate that an error occurred somehow. right now you'd have to look at what it prints since if it times out it just returns and it swallows all exceptions. Let's let this raise, and let's raise on a timeout, and have a higher level format nice error prints and `sys.exit(1)` at least.

---

like much better.